src/parser/jobs/blm/Gauge.js

```diff
@@ -7,210 +7,291 @@ import ACTIONS from 'data/ACTIONS'
-				const offSet = enoRunTime % ENOCHIAN_DURATION_REQUIRED
-				this._enoTimer = event.timestamp - offSet
-				this._poly ++
-				if (this._poly > 1) {
-					this._poly = 1
-					this._overwrittenFoul ++
-				}
+	normalise(events) {
+		// Add initial event
+		this._currentTimestamp = events[0].timestamp
+		this.addEvent()
+
+		for (this._normalizeIndex = 0; this._normalizeIndex < events.length; this._normalizeIndex++) {
+			const event = events[this._normalizeIndex]
+			this._currentTimestamp = event.timestamp
+
+			this.updateStackTimers(event)
+
+			if (!this.parser.byPlayer(event) || !GAUGE_EVENTS.includes(event.type)) {
+				continue
 			}
-		}
 
-		//set _eno to 1 to show it's on. Maybe should have used true/false. Also set timestamp.
-		if (abilityId === ACTIONS.ENOCHIAN.id) {
-			if (!this._eno) {
-				this._eno = true
-				this._enoTimer = event.timestamp
+			switch (event.type) {
+			case 'begincast':
+				break
+			case 'cast':
+				this._onCast(event)
+				break
+			case 'damage':
+				break
+			case 'death':
+				this._onDeath(event)
+				break
 			}
-			this._eno = true
 		}
 
-		//check for AF1 actions and update buffs accordingly. Also check for lost fouls and dropped enos while you're at it.
-		if (AF1_ACTIONS.includes(abilityId)) {
-			if (this._UI > 0) {
-				this.enoDrop()
+		// Add all the events we gathered up in, in order
+		let offset = 0
+		this._toAdd.sort((a, b) => a.insertAfter - b.insertAfter).forEach(event => {
+			events.splice(event.insertAfter + 1 + offset, 0, event)
+			offset++
+		})
 
-			} else {
-				this._AFUITimer = event.timestamp
-				this._AF ++
-				this._AF = Math.min(this._AF, MAX_STACK_COUNT)
+		return events
+	}
+
+	onAstralUmbralTimeout() {
+		this._astralFireStacks = 0
+		this._umbralIceStacks = 0
+		this._astralUmbralStackTimer = 0
+		this.onEnoDropped()
+	}
+
+	onEnoDropped() {
+		if (this._hasEnochian) {
+			if (!this._hasPolyglot) {
+				this._lostFoul++
 			}
+			this._droppedEno++
 		}
+		this._hasEnochian = false
+		this._enochianTimer = 0
+		this._umbralHeartStacks = 0
+		this.addEvent()
+	}
 
-		//check for UI1 actions and update buffs accordingly. Also check for lost fouls and dropped enos while you're at it.
-		if (UI1_ACTIONS.includes(abilityId)) {
-			if (this._AF > 0) {
-				this.enoDrop()
+	onGainPolyglot() {
+		if (this._hasPolyglot) {
+			this._overwrittenFoul++
+		}
+		this._hasPolyglot = true
+		this.addEvent()
+	}
 
-			} else {
-				this._AFUITimer = event.timestamp
-				this._UI ++
-				this._UI = Math.min(this._UI, MAX_STACK_COUNT)
-			}
+	onConsumePolyglot() {
+		if (!this._hasPolyglot && this._overwrittenFoul > 0) {
+			// Safety to catch ordering issues where Foul is used late enough to trigger our overwrite check but happens before Poly actually overwrites
+			this._overwrittenFoul--
 		}
+		this._hasPolyglot = false
+		this.addEvent()
+	}
 
-		//keep track of UH
-		if (abilityId === ACTIONS.BLIZZARD_IV.id) {
-			this._UH = 3
+	onGainUmbralIceStack(event) {
```

Comment:
Lots of repetition in these handlers. Can probably DRY them up a bit by passing stack counts around as args and so on.